snapshotter/pkg/dmsetup/dmsetup.go

```diff
@@ -0,0 +1,310 @@
+		devices = make([]*DeviceInfo, len(lines))
+	)
+
+	for i, line := range lines {
+		var (
+			attr = ""
+			info = &DeviceInfo{}
+		)
+
+		_, err := fmt.Sscan(line,
+			&info.Name,
+			&info.BlockDeviceName,
+			&attr,
+			&info.Major,
+			&info.Minor,
+			&info.OpenCount,
+			&info.TargetCount,
+			&info.EventNumber)
+
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to parse line %q", line)
+		}
+
+		// Parse attributes (see "man dmsetup" for details)
+		info.Suspended = strings.Contains(attr, "s")
+		info.ReadOnly = strings.Contains(attr, "r")
+		info.TableLive = strings.Contains(attr, "L")
+		info.TableInactive = strings.Contains(attr, "I")
+
+		devices[i] = info
+	}
+
+	return devices, nil
+}
+
+// Version returns "dmsetup version" output
+func Version() (string, error) {
+	return dmsetup("version")
+}
+
+// GetFullDevicePath returns full path for the given device name (like "/dev/mapper/name")
+func GetFullDevicePath(deviceName string) string {
+	if strings.HasPrefix(deviceName, DevMapperDir) {
+		return deviceName
+	}
+
+	return DevMapperDir + deviceName
+}
+
+// BlockDeviceSize returns size of block device in bytes
+func BlockDeviceSize(devicePath string) (uint64, error) {
+	data, err := exec.Command("blockdev", "--getsize64", "-q", devicePath).CombinedOutput()
+	output := string(data)
+	if err != nil {
+		return 0, errors.Wrapf(err, output)
+	}
+
+	output = strings.TrimSuffix(output, "\n")
+	return strconv.ParseUint(output, 10, 64)
+}
+
+func dmsetup(args ...string) (string, error) {
+	data, err := exec.Command("dmsetup", args...).CombinedOutput()
+	output := string(data)
+	if err != nil {
+		// Try find Linux error code otherwise return generic error with dmsetup output
+		if errno, ok := tryGetUnixError(output); ok {
+			return "", errno
+		}
+
+		return "", errors.Wrapf(err, "dmsetup %s\nerror: %s\n", strings.Join(args, " "), output)
+	}
+
+	output = strings.TrimSuffix(output, "\n")
+	output = strings.TrimSpace(output)
+
+	return output, nil
+}
+
+// tryGetUnixError tries to find Linux error code from dmsetup output
+func tryGetUnixError(output string) (unix.Errno, bool) {
+	// It's useful to have Linux error codes like EBUSY, EPERM, ..., instead of just text.
+	// Unfortunately there is no better way than extracting/comparing error text.
+	text := parseDmsetupError(output)
+	if text == "" {
+		return 0, false
+	}
+
+	errTableInit.Do(func() {
+		// Precompute map of <text>=<errno> for optimal lookup
+		errTable = make(map[string]unix.Errno)
+		for errno := unix.EPERM; errno <= unix.EHWPOISON; errno++ {
+			errTable[errno.Error()] = errno
+		}
+	})
+
+	err, ok := errTable[text]
+	return err, ok
+}
+
+// dmsetup returns error messages in format:
+// 	device-mapper: message ioctl on <name> failed: File exists\n
+// 	Command failed\n
+// parseDmsetupError extracts text between "failed: " and "\n"
+func parseDmsetupError(output string) string {
+	re := regexp.MustCompilePOSIX("failed: [^:]+$")
```

Comment:
Not super wild about regex on arbitrary strings.  Can we just do a search for "failed: " (`strings.Contains`) and then split at ":" (`strings.SplitN`)?